transport/TcpEndpoint.cpp

```diff
@@ -277,21 +268,54 @@ void TcpEndpoint::continueSend()
     if (_pendingStunRequest && _state == State::CONNECTED)
     {
         // stun requests are always created on own allocator in SendStunRequest
-        nwuint16_t packetLength(_pendingStunRequest->getLength());
-        int rc = _socket.sendAggregate(&packetLength,
-            sizeof(uint16_t),
-            _pendingStunRequest->get(),
-            _pendingStunRequest->getLength());
+        sendPacket(*_pendingStunRequest);
 
         _pendingStunRequest.reset();
-        if (rc != 0)
+    }
+}
+
+void TcpEndpoint::sendPacket(const memory::Packet& packet)
+{
+    if (_remainder.getLength() > 0)
+    {
+        size_t bytesSent;
+        auto rc = _socket.sendAggregate(_remainder.get(), _remainder.getLength(), bytesSent);
+        if (bytesSent < _remainder.getLength())
         {
-            logger::warn("failed send to %s, err (%d) %s",
-                _name.c_str(),
-                _peerPort.toString().c_str(),
-                rc,
-                _socket.explain(rc));
+            memory::Packet tmp;
+            tmp.append(reinterpret_cast<uint8_t*>(_remainder.get()) + bytesSent, _remainder.getLength() - bytesSent);
+            tmp.copyTo(_remainder);
+        }
+        else
+        {
+            _remainder.setLength(0);
         }
+        if (_remainder.getLength() > 0)
```

Comment:
isn't this `if` the same of line `283`( `if (bytesSent < _remainder.getLength())`)? Maybe you can lof the `warn` and `return` on the first `if`